presto-test-jdbc-compatibility/src/test/java/io/prestosql/TestJdbcCompatibility.java

```diff
@@ -0,0 +1,479 @@
+        }
+
+        assertThat(driverVersion())
+                .isEqualTo(VERSION_UNDER_TEST.get());
+    }
+
+    @Test
+    public void ensureUsingProperCliVersion()
+            throws ClassNotFoundException
+    {
+        if (VERSION_UNDER_TEST.isEmpty()) {
+            throw new SkipException("PRESTO_JDBC_VERSION environment variable is missing");
+        }
+
+        assertThat(Class.forName("io.prestosql.jdbc.$internal.client.StatementClientFactory").getPackage().getImplementationVersion())
+                .isEqualTo(Integer.toString(VERSION_UNDER_TEST.get()));
+    }
+
+    @Test
+    public void testLongPreparedStatement()
+            throws Exception
+    {
+        String sql = format("SELECT '%s' = '%s'", repeat("x", 100_000), repeat("y", 100_000));
+
+        try (ResultSet rs = runQuery(sql)) {
+            assertThat(rs.next()).isTrue();
+            assertThat(rs.getBoolean(1)).isFalse();
+            assertThat(rs.next()).isFalse();
+        }
+    }
+
+    @Test
+    public void testTimestampSelect()
+    {
+        String query = "SELECT TIMESTAMP '2012-10-31 01:00'";
+        checkRepresentation(query, Timestamp.valueOf("2012-10-31 01:00:00.000"), ResultSet::getTimestamp);
+    }
+
+    @Test
+    public void testTimestampWithTimeZoneSelect()
+    {
+        if (hasBrokenParametricTimestampWithTimeZoneSupport()) {
+            throw new SkipException("This version raports PARAMETRIC_DATETIME client capability but TIMESTAMP WITH TIME ZONE is not supported");
+        }
+
+        String query = format("SELECT timestamp '2012-10-31 01:00 %s'", TESTED_TZ);
+        checkRepresentation(query, Timestamp.valueOf("2012-10-30 10:15:00.000"), ResultSet::getTimestamp);
+    }
+
+    @Test
+    public void testSelectParametricTimestamp()
+    {
+        String timestamp = "2421-01-19 15:55:23.393456298901";
+
+        for (int precision = 0; precision <= TIMESTAMP_MAX_PRECISION; precision++) {
+            String query = format("SELECT cast(TIMESTAMP '%s' as TIMESTAMP(%d))", timestamp, precision);
+            Timestamp expectedTimestamp = roundedSqlTimestamp(timestamp, precision, TIMESTAMP_JDBC_MAX_PRECISION, false);
+
+            checkRepresentation(query, expectedTimestamp, ResultSet::getTimestamp);
+            checkRepresentation(query, expectedTimestamp, ResultSet::getObject);
+            checkRepresentation(query, roundedTimestamp(timestamp, precision, TIMESTAMP_MAX_PRECISION, false), ResultSet::getString);
+        }
+    }
+
+    @Test
+    public void testSelectParametricTimestampWithTimeZone()
+    {
+        if (hasBrokenParametricTimestampWithTimeZoneSupport()) {
+            throw new SkipException("This version raports PARAMETRIC_DATETIME client capability but TIMESTAMP WITH TIME ZONE is not supported");
+        }
+
+        String timestamp = "2020-07-16 15:55:23.383345789012";
+        String localTimestamp = "2020-07-16 02:10:23.383345789012";
```

Comment:
Having 2 different values like this looks like we have a bug (which we do not have).
Let's derived one from the other

```
Timestamp expected = new Timestamp(LocalDateTime.parse(timestamp)
                .atZone(ZoneId.of(TESTED_TZ))
                .toInstant().toEpochMilli());
```

Reply:
We have 12 digits of precision here and Timestamp holds only 9